crates/binjs_io/src/entropy/predict.rs

```diff
@@ -8,128 +10,321 @@ use std::rc::Rc;
+        }
+    }
+
+    // Methods that make sense only while we are collecting instances.
+
+    impl<NodeValue> ContextInformation<NodeValue, usize> where NodeValue: Eq + Hash {
+        /// Register a value as being used in this context.
+        pub fn add(&mut self, node_value: NodeValue) {
+            self.by_value.entry(node_value)
+            .and_modify(|instances| *instances += 1)
+            .or_insert(1);
+        }
+    }
+
+    impl<NodeValue> ::entropy::probabilities::InstancesToProbabilities for ContextInformation<NodeValue, usize> where NodeValue: Clone + Eq + Hash {
+        type AsProbabilities = ContextInformation<NodeValue, SymbolInfo>;
+        fn instances_to_probabilities(self, _description: &str) -> ContextInformation<NodeValue, SymbolInfo> {
+            let instances: Vec<_> = self
+                .by_value
+                .values()
+                .map(|x| *x as u32)
+                .collect();
+
+            let distribution = std::rc::Rc::new(std::cell::RefCell::new(range_encoding::CumulativeDistributionFrequency::new(instances).
+                unwrap()));
+            // FIXME: This will fail if `by_value` is empty.
+            // FIXME: We should have a fallback distribution in case everything is empty.
+
+            let (by_value, by_index): (HashMap<_, _>, Vec<_>) = self.by_value
+                .into_iter()
+                .enumerate()
+                .map(|(index, (value, _))| {
+                    let for_by_value = (value.clone(), SymbolInfo {
+                        index,
+                        distribution: distribution.clone(),
+                    });
+                    let for_by_index = value;
+                    (for_by_value, for_by_index)
+                })
+                .unzip();
+            ContextInformation {
+                by_value,
+                by_index
+            }
+        }
+    }
 }
+use self::context_information::ContextInformation;
 
-/// A generic predictor, associating a context and a key to a value.
-#[derive(Debug, Default, Serialize, Deserialize)]
-pub struct ContextPredict<C, K, T> where C: Eq + Hash + Clone, K: Eq + Hash + Clone {
-    by_context: HashMap<C, HashMap<K, T>>,
+/// A generic mechanism used to predict possible values in a given context (e.g.
+/// AST path or file position) in a file.
+///
+/// This mechanism is meant to be used as follows:
+///
+/// 1. Create a new `ContextPredict<Context, NodeValue, usize>` and use `ContextPredict::add` to count the number
+///     of instances of each possible `NodeValue` in each possible `Context`. Alternatively, load this data
+///     from an existing dictionary.
+/// 2. Convert the `ContextPredict<Context, NodeValue, usize>` into a `ContextPredict<Context, NodeValue, SymbolInfo>`
+///     by calling `ContextPredict::instances_to_probabilities`.
+/// 3. Use method `ContextPredict::<_, _, SymbolInfo>::by_node_value` and `by_node_value_mut` to get the statistics
```

Comment:
method names don't match to the actual method names.
but again, please please rename them to include prefix "what the value of the container is"